<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" dir="ltr" lang="en">
<head>
	<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
	<meta name="keywords" content="Problem with AFS for tdilc4202m000, baan,baanerp,erp,forum,discussion,bulletin board" />
	<meta name="description" content="[Archive] Problem with AFS for tdilc4202m000 AFS/DDC/OLE: Function servers" />
	
	<title>Problem with AFS for tdilc4202m000 [Archive]  - Baanboard.com</title>
	<link rel="stylesheet" type="text/css" href="../styles.css" />
</head>
<body>
<div class="pagebody">
<div id="navbar"><a href="index.html">Newbaanboard</a> &gt; <a href="index.html">Baan SIGs</a> &gt; <a href="index.html">AFS/DDC/OLE: Function servers</a> &gt; Problem with AFS for tdilc4202m000</div>
<hr />
<div class="pda"></div>

<hr />

<div class="post"><div class="posttop"><div class="username">dkoester</div><div class="date">1st August 2011, 13:02</div></div><div class="posttext">Hello,<br />
<br />
I have a spooky problem with an afs-function for a standard session (tdilc4202m000).<br />
<br />
For background information:<br />
We would like to optimize our &quot;sales/delivery-process&quot; and therefore I have to automate some baan-standard sessions, so that they will be executed one after another (in an own session).<br />
<br />
It should work as follows:<br />
1) create outbound data for one or some position(s) of a sales order (tdilc4201m000)<br />
2) create picking list (not the baan-standard; it is an own development, that only saves the picking number in the field &quot;tdsls045.pino&quot;)<br />
3) release outbound data for the combination sales order (orno) / picking list (pino) (tdilc4202m000)<br />
4) deliver cost items with Session tdsls4120m000<br />
5) create delivery note for the combination sales order (orno) / picking list (pino) (tdsls4403m000) <br />
6) create invoice for sales order (tdsls4404m000) <br />
<br />
This six steps/sessions should be handled complete by AFS. All works fine, except the AFS-call of the session tdilc4202m000 (release outbound data). In some cases the session only returns the error &quot;no data found for report, no report is printed (message code: tccoms0024)&quot;, although the session got the right values and the tdilc401 has &quot;unreleased records&quot; for the corresponding order position. Executing the session manually with the same values works without an error ...<br />
<br />
Code for the AFS-function of tdilc4202m000:<br />
<br />
function extern tdlubilc401.release.outbound.data.for.pino(domain tcmcs.st12 runnumber, 							 <br />
							domain	tcorno	   orno,							<br />
							domain	tcpino	   pino,<br />
							domain	tcyesno	   deliver.costitem,<br />
							domain	tilub.devc device, <br />
							ref domain tcmcs.str14 	msg.code,<br />
							ref domain tcmcs.str215 error.msg)<br />
{<br />
DLLUSAGE<br />
	Function to release outbound data (sls)<br />
ENDDLLUSAGE<br />
	<br />
	sessioncode = &quot;tdilc4202m000&quot;<br />
<br />
	if deliver.costitem = empty then<br />
		deliver.costitem = tcyesno.no<br />
	endif<br />
<br />
	stpapi.put.field(sessioncode, &quot;runnumber.f&quot;, str$(runnumber))<br />
	stpapi.put.field(sessioncode, &quot;runnumber.t&quot;, str$(runnumber))<br />
	stpapi.put.field(sessioncode, &quot;koor.f&quot;, str$(tdilc.koob.act.sls))<br />
	stpapi.put.field(sessioncode, &quot;koor.t&quot;, str$(tdilc.koob.act.sls))<br />
	stpapi.put.field(sessioncode, &quot;cuno.f&quot;, &quot;&quot;)<br />
	stpapi.put.field(sessioncode, &quot;cuno.t&quot;, &quot;ZZZZZZ&quot;)		<br />
	stpapi.put.field(sessioncode, &quot;orno.f&quot;, str$(orno))<br />
	stpapi.put.field(sessioncode, &quot;orno.t&quot;, str$(orno))<br />
	stpapi.put.field(sessioncode, &quot;pino.f&quot;, str$(pino))<br />
	stpapi.put.field(sessioncode, &quot;pino.t&quot;, str$(pino))	<br />
	stpapi.put.field(sessioncode, &quot;opno.f&quot;, &quot;0&quot;)<br />
	stpapi.put.field(sessioncode, &quot;opno.t&quot;, &quot;999&quot;)	<br />
	stpapi.put.field(sessioncode, &quot;prdt.f&quot;, &quot;&quot;)<br />
	stpapi.put.field(sessioncode, &quot;prdt.t&quot;, str$(date.num()+500))<br />
	stpapi.put.field(sessioncode, &quot;ddat.f&quot;, str$(date.num()-500))<br />
	stpapi.put.field(sessioncode, &quot;ddat.t&quot;, str$(date.num()+500))<br />
	stpapi.put.field(sessioncode, &quot;cprj.f&quot;, &quot;&quot;)<br />
	stpapi.put.field(sessioncode, &quot;cprj.t&quot;, &quot;ZZZZZZ&quot;)<br />
	stpapi.put.field(sessioncode, &quot;item.f&quot;, &quot;&quot;)<br />
	stpapi.put.field(sessioncode, &quot;item.t&quot;, &quot;ZZZZZZZZZZZZZZZZ&quot;)<br />
	stpapi.put.field(sessioncode, &quot;cntr.f&quot;, &quot;&quot;)<br />
	stpapi.put.field(sessioncode, &quot;cntr.t&quot;, &quot;ZZZ&quot;)<br />
	stpapi.put.field(sessioncode, &quot;cwar.f&quot;, &quot;&quot;)<br />
	stpapi.put.field(sessioncode, &quot;cwar.t&quot;, &quot;ZZZ&quot;)<br />
	stpapi.put.field(sessioncode, &quot;crte.f&quot;, &quot;&quot;)<br />
	stpapi.put.field(sessioncode, &quot;crte.t&quot;, &quot;ZZZZZ&quot;)<br />
	stpapi.put.field(sessioncode, &quot;deliver.costitem&quot;, str$(etol(deliver.costitem)))<br />
	stpapi.put.field(sessioncode, &quot;prnt.error.report&quot;, str$(etol(tcyesno.no)))<br />
	stpapi.set.report(sessioncode, &quot;rtdilc420201000&quot;, device, error.msg)<br />
	if isspace(error.msg) then<br />
		stpapi.continue.process(sessioncode, error.msg)<br />
		msg.code = stpapi.get.mess.code(sessioncode, error.msg)<br />
	else<br />
		msg.code = stpapi.get.mess.code(sessioncode, error.msg)<br />
	endif<br />
	stpapi.end.session(sessioncode)	<br />
}<br />
<br />
<br />
Hint: The function will be only executed when the corresponding outbound data is not yet released. (query of tdilc401.proc)<br />
<br />
Unfortunately I have no explanation for this &quot;error&quot; as it seems to me that there is no error. According to a created logfile (afs.log) the session will be called correctly, but inexplicably returns the mentioned error, although there is data for continuing. <br />
<br />
I hope I haven't violated rules (in my first post) and someone can help me.<br />
<br />
Thanks for your time in advance.<br />
<br />
Regards,<br />
Dennis Koester<br />
<br />
P.S. Our baan-system is baan4 c4 on hp-ux 11i v3 (B.11.31)</div></div><hr />


<div class="post"><div class="posttop"><div class="username">mark_h</div><div class="date">1st August 2011, 15:47</div></div><div class="posttext">Have you confirmed the fields?  I did not see anything right off the top that is incorrect. Can you duplicate the problem?  If so try this - run the release code a second time.  I am wondering if you are running these routines one after another and in some cases the previous routine, which I assume generates the outbound, might not have completed in the background.  I have had to use suspend before to slow things up a little.  Not sure this is the case.</div></div><hr />


<div class="post"><div class="posttop"><div class="username">dkoester</div><div class="date">2nd August 2011, 09:12</div></div><div class="posttext">Good morning mark,<br />
<br />
thanks for the fast reply.<br />
<br />
Yes I've confirmed the field values by debugging the scripts and logging the afs-function (command &quot;-set AFSLOG=1&quot; in the bw configuration) with checking the created file for a wrong given value. Could not find anything conspicuous ...<br />
<br />
Interestingly the error occurs irregular that means that I had for example different situations:<br />
<br />
1) 1 process*: ok<br />
2) 1 process: error<br />
3) 2 processes: first ok, second error (and vice versa)<br />
4) 3 processes: first ok, second and third error (and different constellation)<br />
*process: a try to release outbound data for the combination orno/pino<br />
<br />
Your idea to use the suspend command to &quot;slow down the script&quot; sounds not bad. It's still worth a try. :) I'm even pretty sure that I didn't get an error when I debugged the whole background session including the afs-library ... and that would confirm your assumption.<br />
<br />
The automation of the session for generating outbound data (tdilc4201m000) didn't return any errors till now, so that the tdilc401 is always filled correctly.<br />
<br />
Unfortunately I forgot to mention that the AFS-functions for &quot;generating outbound data&quot; and &quot;creating a picking list&quot; (which based on the created outbound data!) is called at a different time. We need this data for a transfer in an &quot;external picking-system&quot; where the data will be processed and returned to baan. That only for background. ;)<br />
<br />
I've already changed the printer and reset the user defaults (&quot;ttstpdeldeflt&quot; for root-account, which starts our background-jobs) too, just to exclude it.<br />
<br />
Greetings,<br />
Dennis</div></div><hr />


<div class="post"><div class="posttop"><div class="username">mark_h</div><div class="date">2nd August 2011, 14:52</div></div><div class="posttext">I forgot about user defaults - one of the reasons I always try to make sure I put all the fields.  On our system we do generate and release back to back in a batch job - this is all production orders.  I know at one point in time a couple of our AFS routines would error out after processing x number of records - x varied.  When that happened we would just restart them where they left off(this was by hand not in batch).  The issue slowly went away as we applied SP's.</div></div><hr />


<div class="post"><div class="posttop"><div class="username">dkoester</div><div class="date">3rd August 2011, 10:39</div></div><div class="posttext">Last week I have experienced that user-defaults will be queried/used although all form fields were filled with values, i.e. I wanted to generate an invoice (with AFS), but got the error that &quot;the invoice date is not correct&quot; (or s.th. like that), cause I saved an old date from 2009. No idea why, but after deleting the user-defaults for this session, it worked. :)<br />
<br />
For our production we developed a session a long time ago, where an user can create or release outbound data for one entered production order by using a button. That works till now without any problems and the call/use of AFS is nearly the same, the only difference is just the order type and of course the runnumber. But the runnumber is only for further information or filtering isn't it?<br />
<br />
Because of the effect that even one automatic release-try (in the background-job) can return an error, it might be not very easy to create a workaround so that the process won't be interrupted by an error. By the way in a second (equal) baan-system (for &quot;testing&quot;, devolpment) the background session throws no errors. Perhaps I should mention it too. :rolleyes:<br />
<br />
What do you mean with &quot;applied SP's&quot;?</div></div><hr />


<div class="post"><div class="posttop"><div class="username">mark_h</div><div class="date">3rd August 2011, 18:08</div></div><div class="posttext">Last week I have experienced that user-defaults will be queried/used although all form fields were filled with values, i.e. I wanted to generate an invoice (with AFS), but got the error that &quot;the invoice date is not correct&quot; (or s.th. like that), cause I saved an old date from 2009. No idea why, but after deleting the user-defaults for this session, it worked. :)<br />
<br />
I hate saving defaults in some cases.  I have turned this option off on a few sessions.<br />
<br />
<br />
 That works till now without any problems and the call/use of AFS is nearly the same, the only difference is just the order type and of course the runnumber. But the runnumber is only for further information or filtering isn't it?<br />
<br />
The order type could be activating something in the code that is throwing the error.  I don't see how it could be runnumber.<br />
<br />
<br />
Because of the effect that even one automatic release-try (in the background-job) can return an error, it might be not very easy to create a workaround so that the process won't be interrupted by an error. By the way in a second (equal) baan-system (for &quot;testing&quot;, devolpment) the background session throws no errors. Perhaps I should mention it too. :rolleyes:<br />
<br />
What do you mean with &quot;applied SP's&quot;?<br />
It means service packs - this problem occured on an early service pack.  We are now on something like Service pack 28(I think) and the issue is gone.  What this tells me (1 system works and 1 does not) is that they are not equal somewhere.  It could be porting set, STPAPI libraries, service pack release or even a 1 off patch on the session.  The first this I would check is the session if possible.  Then I would get the latest stpapi patches or libraries available.  And yes even the porting set might need to updated to fix the problem.</div></div><hr />


<div class="post"><div class="posttop"><div class="username">dkoester</div><div class="date">5th August 2011, 11:49</div></div><div class="posttext">hello mark,<br />
<br />
sry for the late feedback, i'm very stressed with different Problems at the moment.<br />
<br />
The order type could be activating something in the code that is throwing the error. I don't see how it could be runnumber.<br />
<br />
I'll try to check it. But in theory every AFS call (for the same combination orno/pino), which throws once an error should throw an error again or? On Wednesday I implemented your proposal with the &quot;suspend-command&quot; (with 1000 ms) and then the background job could released two picking lists without problems. I already thought that your suggestion was the solution but yesterday I got the same error again ... <br />
As an workaround I created a Session where a user can choose one picking list (which got an afs error) and start another try. In the session I use the same afs call of course. It worked. :rolleyes:<br />
<br />
It means service packs - this problem occured on an early service pack. We are now on something like Service pack 28(I think) and the issue is gone. What this tells me (1 system works and 1 does not) is that they are not equal somewhere. It could be porting set, STPAPI libraries, service pack release or even a 1 off patch on the session. The first this I would check is the session if possible. Then I would get the latest stpapi patches or libraries available. And yes even the porting set might need to updated to fix the problem.<br />
<br />
I have talked about the situation with our system administrator, cause I'm &quot;only the&quot; developer. :) Currently sp22 is installed and because of no critical problems, which would &quot;justify&quot; a newer SP, we didn't touch the system since the end of 2007. Due to the fact that the second system, our test-system, is a restore of the live-system they are technically identical except some new developments. Both system have the porting set 6.1c.09 (one with PA.3846, the other with PA.3847 because of different hardware) installed. Therefore the stpapi libraries must be equal too. But on the other hand something must be different what explains that the error only occur in the live-system. What about permissions? The background job will be executed by a root user (system-user: bsp).<br />
<br />
greetings,<br />
Dennis<br />
<br />
edit: I forgot to mention that I've checked the session tdilc4202m000 with the session &quot;ttstpsessinfo&quot;. At first glance I saw no difference.</div></div><hr />


<div class="post"><div class="posttop"><div class="username">mark_h</div><div class="date">8th August 2011, 15:38</div></div><div class="posttext">Well you would think the same record would error out, but like I mentioned that one problem we had would process a random number of records before the error would show up.  I always thought it was something in the background that was not getting cleaned up correctly.  One of the reasons I almost always start and end a session for each record.<br />
<br />
Yes - system admins. :)  Yes - it could be permissions, but I just can't imagine where.  I mean one works and then the next doesn't.  It seems to me it could be system related, but I have no clue as to where.  Just too weird and I can't think anything to true.</div></div><hr />


<div class="post"><div class="posttop"><div class="username">dkoester</div><div class="date">9th August 2011, 08:30</div></div><div class="posttext">ok, if I understand you correctly you recommend to change the script/session, so that only one record will be found and processed. After that the session should be closed and then I've to start the session again to find the next record. Right?<br />
<br />
greetings<br />
Dennis</div></div><hr />


<div class="post"><div class="posttop"><div class="username">mark_h</div><div class="date">9th August 2011, 15:41</div></div><div class="posttext">ok, if I understand you correctly you recommend to change the script/session, so that only one record will be found and processed. After that the session should be closed and then I've to start the session again to find the next record. Right?<br />
<br />
greetings<br />
Dennis<br />
<br />
It is just something I do.  So if I have a file with 200 parts to transfer from one bin to another.  I open inventory session, transfer part, close inventory session, get next part - start again.  This does add overhead, but my own opinion is that it makes it cleaner(nothing floating around in the background that might confuse things).  So for out bound they can only process one order at a time in my session - so basically they enter the order, then it generate, then it releases. And I always close all the sessions after they run. I have just followed this from early on - it helped when doing inventory adjustments on(we have to add pegging data) our A&amp;D system.  That was the killer that would abort with some generic error - by closing things it seemed to process more records before this abort.<br />
<br />
I can only think of a couple places where I process more records leaving sessions open.  I know 2 of them are delete grp purchase and grp production orders.  Those sessions might process 10 or 1000 records before one errors out.  These sessions only IT people run.  We have to cleanup orders every so often(once a quarter) or GRP starts taking days to run.<br />
<br />
Come to think of it I have had somethings work in dev/test and not in production.  Including AFS code - usually I debugged them in production and the new code worked in test.  ANother aggrevating error still occurs(as far as I know) - in dev/test I can take a double and say &quot;if x&gt;0 then&quot;.  It works there, but move the code to production and it will not work consistently.  We had several programs fail in production that were running okay for years.  And of course - nobody did anything to cause this.  I had to change several programs use double.cmp() to get them working again.</div></div><hr />


<div class="post"><div class="posttop"><div class="username">dkoester</div><div class="date">11th August 2011, 12:00</div></div><div class="posttext">hey mark,<br />
<br />
It is just something I do. So if I have a file with 200 parts to transfer from one bin to another. I open inventory session, transfer part, close inventory session, get next part - start again. This does add overhead, but my own opinion is that it makes it cleaner(nothing floating around in the background that might confuse things). So for out bound they can only process one order at a time in my session - so basically they enter the order, then it generate, then it releases. And I always close all the sessions after they run. I have just followed this from early on - it helped when doing inventory adjustments on(we have to add pegging data) our A&amp;D system. That was the killer that would abort with some generic error - by closing things it seemed to process more records before this abort.<br />
Your approach sounds logical and sensible. I prefer the same way but I always thought that the first call of<br />
stpapi.put.field(sessioncode, sessionfield, value)<br />
and the call (at the end of AFS) of<br />
stpapi.end.session(sessioncode)<br />
would exactly do what you mentioned: <br />
<br />
 open session<br />
 fill the fields with data<br />
 press button(s)<br />
 close the session<br />
<br />
and that for each case, just as I would use the session manually. But for testing I have changed the script, so that only one picking list will be processed in the whole session. To handle more than one list, I've edited the background job and insert some copies of the session. The result should be rather the same ... but unfortunately not. I still got an error. <br />
<br />
Before I tried it again, I looked for equal calls to release outbound for production orders in other (and older) scripts. I found s.th. I can not explain, but it works:<br />
stpapi.put.field(sessioncode, &quot;runnumber&quot;, str$(runnumber))<br />
The field doesn't exist, only &quot;runnumber.f&quot; and &quot;runnumber.t&quot; ... The session is scary. <br />
<br />
Come to think of it I have had somethings work in dev/test and not in production. Including AFS code - usually I debugged them in production and the new code worked in test. ANother aggrevating error still occurs(as far as I know) - in dev/test I can take a double and say &quot;if x&gt;0 then&quot;. It works there, but move the code to production and it will not work consistently. We had several programs fail in production that were running okay for years. And of course - nobody did anything to cause this. I had to change several programs use double.cmp() to get them working again.<br />
Very interesting hint. Fortunately I was not confronted with such a situation till now, and I really hope that not &quot;s.th. stupid&quot; is the cause for the irregular errors.<br />
<br />
Actual I put my hope on Infor respectively the support-case our external support has opened. I'll get slowly crazy with the situation. :(<br />
<br />
greetings<br />
Dennis</div></div><hr />


<div class="post"><div class="posttop"><div class="username">mark_h</div><div class="date">11th August 2011, 20:35</div></div><div class="posttext">Glad you found the solution.  I believe if you search around you will see some boi.call variables that have to be set for some sessions to work.  This process can be tough to figure out solutions.</div></div><hr />



</div>
</body>
</html>